components/slack/slack.app.js

```diff
@@ -3,6 +3,243 @@ const { WebClient } = require('@slack/web-api')
 module.exports = {
   type: "app",
   app: "slack",
+  propDefinitions: {
+    publicChannel: {
+      type: "string",
+      label: "Channel",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["public_channel"]
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            if (c.is_im) {
+              return { label: `Direct messaging with: @${userNames[c.user]}`, value: c.id }
+            } else if (c.is_mpim) {
+              return { label: c.purpose.value, value: c.id }
+            } else {
+              return { label: `${c.name}`, value: c.id }
+            }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    privateChannel: {
+      type: "string",
+      label: "Channel",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["private_channel"]
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            return { label: `${c.name}`, value: c.id }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    user: {
+      type: "string",
+      label: "User",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["im"]
+          // TODO use paging
+          userNames = {}
+          for (const user of (await this.users()).users) {
+            userNames[user.id] = user.name
+          }
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
```

Comment:
So it seems like this will only get the list of conversations for users you've ever had a conversation with?  I would think a prop named 'user' would return user objects.  In fact I think the action should be using that prop (one that return all the users), and then use postMessage with the user id.  (Though I also feel like that might not work?)